libs/core/iterator_support/include/hpx/iterator_support/boost_iterator_categories.hpp

```diff
@@ -0,0 +1,214 @@
+//  Copyright (c) 2021 Hartmut Kaiser
+//
+//  SPDX-License-Identifier: BSL-1.0
+//  Distributed under the Boost Software License, Version 1.0. (See accompanying
+//  file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+
+// (C) Copyright Jeremy Siek 2002.
+
+#pragma once
+
+#include <hpx/config.hpp>
+
+// we deliberately leave the header guards in place to protect against
+// #including this header and the original one from Boost at the same time
+#ifndef BOOST_ITERATOR_CATEGORIES_HPP
+#define BOOST_ITERATOR_CATEGORIES_HPP
```

Comment:
Hmm, looks like my comment was lost... I'll try to summarize: I don't love this, but this header would be the safest one to do this with since it hasn't changed in the last 4 years in Boost. I'm wondering if we could do the same here as what we've done with support for serialization of Boost types? That way when/if all remaining Boost compatibility is behind CMake options they could just be consolidated into a single `HPX_WITH_BOOST_COMPATIBILITY/SUPPORT` option or similar. What do you think?

Reply:
That would work. Let me see what I can do. I think that using this option instead of the unwieldy header guard would be a possible solution (even if it isn't perfect either).

Reply:
I introduced a compatibility option that I left `OFF` by default. So it shouldn't be a problem for people enable compatibility if needed.

Reply:
Actually I had to enable it by default as the boost header is indirectly included somewhere.